benchmarks/src/main/scala/monix/TaskGatherUnordered.scala

```
@@ -0,0 +1,186 @@
+  var taskSize: Int = _
+
+  @Param(Array("10", "500", "10000"))
+  var collSize: Int = _
+
+
+  def sum(n: Int, acc: Long = 0): Task[Long] =
+    if (n == 0) Task.evalAlways(acc)
+    else Task.evalAlways(n).flatMap(x => sum(x-1, acc + x))
+
+  def tasks[M[_] <: TraversableOnce[_]](implicit
+    cbf: CanBuildFrom[Nothing, Task[Long], M[Task[Long]]]
+  ): M[Task[Long]] = {
+    val builder = cbf.apply()
+    builder.sizeHint(collSize)
+    for (_ <- 0 until collSize) { builder += sum(taskSize) }
+    builder.result()
+  }
+
+
+  @Benchmark
+  def gather_list(): List[Long] = {
+    val task = Task.gather(tasks[List])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def gather_vector(): Vector[Long] = {
+    val task = Task.gather(tasks[Vector])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_blocking_list(): List[Long] = {
+    val task = gatherUnorderedBlocking(tasks[List])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_blocking_vector(): Vector[Long] = {
+    val task = gatherUnorderedBlocking(tasks[Vector])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_nonblocking_list(): List[Long] = {
+    val task = gatherUnorderedNonBlocking(tasks[List])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_nonblocking_vector(): List[Long] = {
+    val task = gatherUnorderedNonBlocking(tasks[Vector])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+
+  def gatherUnorderedBlocking[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])(
+    implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]
+  ): Task[M[A]] = {
+    Task.unsafeCreate { (scheduler, conn, finalCallback) =>
+      val lock = new Object()
+      val builder = cbf(in)
+      var remaining = in.size
+
+      val composite = CompositeCancelable()
+      conn.push(composite)
+
+      for (task <- in; if !composite.isCanceled) {
```

In this implementation, using `map` (`yield`) instead of `foreach` makes the benchmark hang forever and the inner block never runs. Do you have any clue? Maybe it's throwing an exception that is just swallowed somewhere...